csrc/ops/arith.cpp

```diff
@@ -1492,6 +1492,17 @@ TensorView* expand(TensorView* inp, const std::vector<Val*>& expanded_sizes) {
       // Note that expansion to 1 just means its extent becomes 1 and
       // does not mean the ID becomes a broadcast.
       out_id_builder.extent(maybeCastOp(DataType::Index, expanded_sizes[i]));
+    } else if (
```

Comment:
IIRC, the dynamic reshape API is added by @jacobhinkle

Does this change look reasonable to you?

Reply:
This is good for now since it fixes, but we should probably file an issue with a test that has dynamic shapes like 
```c++
  auto tv0 = makeSymbolicTensor(2);
  fusion->addInput(tv0);

  auto s0 = IrBuilder::create<Val>(DataType::Index);
  auto s1 = IrBuilder::create<Val>(DataType::Index);
  auto s2 = IrBuilder::create<Val>(DataType::Index);
  fusion->addInput(s0);
  fusion->addInput(s1);
  fusion->addInput(s2);
  fusion->addInput(s3);

  auto tv1 = reshape(tv0, {s0, s1});
  auto tv2 = expand(tv1, {s0, s2});

  fusion->addOutput(tv2);
```
The long-term solution would be to accept `Symbolic` inputs here in the definition op and also add an implementation for the method `ExpandOp::checkConcretization(Val* old_val, Val* new_val)` which checks that `old_val` is an `IterDomain` in the input that satisfies the necessary conditions.

Reply:
Why do we need to create a symbolic IterDomain with a concrete size of 1?

Reply:
Opened #1398 .

Would appreciate a review as well :stuck_out_tongue_winking_eye:

Reply:
> Why do we need to create a symbolic IterDomain with a concrete size of 1 in the first place?

That's a good question. We can probably do a clean up on the `reshape` logic when new sizes are just static.

Reply:
> That's a good question. We can probably do a clean up on the reshape logic when new sizes are just static.

Yeah, I think we could just create a broadcast domain if the extent is known to be 1, although this only matters with static reshape, so not a common case. The more fundamental fix would be to support symbolic domains with expand, as Jacob mentioned.

Reply:
Yeah, I think it's fine to create it as `Broadcast` for constant 1. We just need to be sure _some_ `IterDomain`s are `Symbolic` in the output in order to detect a dynamic reshape, but these broadcasts should be safe I think.

Reply:
> We just need to be sure some IterDomains are Symbolic in the output in order to detect a dynamic reshape

Yeah, that's what I also suspected. This would not work if all the shape extents of a reshape op are static. We would not need to create symbolic domains since we knew the actual sizes, but then we would miss the reshape would need to be concretized. I guess we would need to have a different algorithm to detect such a case, which shouldn't be too much of an issue, but again since this is a special case, I don't think we would need to prioritize.

Reply:
> We just need to be sure some IterDomains are Symbolic in the output in order to detect a dynamic reshape

> This would not work if all the shape extents of a reshape op are static. 

Thanks. I tried to special case on the constant 1 dimension for reshape. I think it's safe, as long as we have "some other scalar in the reshape that's not constant 1", we'll always have a Symbolic Iterdomain on the output, so a dynamic reshape will be detected.

Reply:
Meanwhile, I'm tracking the general dynamic reshape + broadcast issue here: https://github.com/NVIDIA/Fuser/issues/1398

Reply:
I thought about this earlier... if we are reshaping a tensor to `[1, 1, 1, ..., 1]`, that implies the input tensor would also needs to be in a `compatible` shape. So it needs to be a size-1 tensor anyway.
Given that we specialize tensor dimension with size-1 as broadcast, the extent would be a constant and the corresponding reshape would become a static reshape. https://github.com/NVIDIA/Fuser/blob/61092461a6abc66746947d98ede565d090e28267/csrc/ops/alias.cpp#L115-L118

Let me know if I missed anything.

Reply:
> symbolic input domains are indeed just broadcast domains,

I *think* the goal is to mark any `static symbolic` domain that is a broadcast **as a broadcast domain**, instead of a symbolic domain.
So this PR only handles reshape. If we apply similar changes to `pad` & `slice`, we should be alright?

Let me see if I can make up some script to break `tryStaticReshape`.

If we are generally ok with the approach we are taking, do you mind if we follow up other operations with symbolic domains in separate issues? just so we can unblock the failing cases for now.

> I think this propagation of static shapes is possible and preferrable, but it might not be that trivial.

yeah going forward there will be more work needed to patch other operations that could give us symbolic domains.
We still need to support `dynamic` symbolic domains, once that is in, this `static` symbolic domain special handling for broadcasting becomes only an "optimization"?

Reply:
I was just referring to a symbolic IterDomain where the extent is just a constant.